### PR TITLE
Issues 2044 2045 2046

### PR DESCRIPTION
**Pull request by JustinKyleJames**
Please take a look at this pull request for issues irods/irods_resource_plugin_s3#2044 irods/irods_resource_plugin_s3#2045 and irods/irods_resource_plugin_s3#2046.

### CODE REVIEW

**Review by alanking — CHANGES_REQUESTED** [review]
Overall, the change looks good. Had a few things to point out.

File: s3/libirods_s3.cpp [hunk 1]
```diff
@@ -101,12 +102,12 @@
 // Pairing this with LIBS3 error injection will exercise the error recovery
 // and retry code paths.
 static boost::mutex g_error_mutex;
-static long g_werr{0}; // counter
-static long g_rerr{0}; // counter
-static long g_merr{0}; // counter
-static const long g_werr_idx{4}; // Which # pwrite to fail
-static const long g_rerr_idx{4}; // Which # pread to fail
-static const long g_merr_idx{4}; // Which part of Multipart Finish XML to fail
+static int64_t g_werr{0}; // counter
```
Comment (korydraughn) [inline s3/libirods_s3.cpp:105]: Consider adding the `std::` prefix on the fixed-wdith integer types.

Although they aren't needed, including the namespace prefix future proofs us from changes to the C++ standard.

File: s3/libirods_s3.cpp [hunk 2]
```diff
@@ -101,12 +101,12 @@
...
+static int64_t g_rerr{0}; // counter
+static int64_t g_merr{0}; // counter
+static const int64_t g_werr_idx{4}; // Which # pwrite to fail
+static const int64_t g_rerr_idx{4}; // Which # pread to fail
+static const int64_t g_merr_idx{4}; // Which part of Multipart Finish XML to fail
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply s3/libirods_s3.cpp:110]: I agree with your assessment :)

File: s3/libirods_s3.cpp [hunk 3]
```diff
@@ -589,15 +592,14 @@ irods::error s3InitPerOperation (
...
+        auto msg = fmt::format("[resource_name={}]  - Error initializing the S3 library. Status = {}.",
+                resource_name, status);
+
+        if(status >= 0) {
+            msg += fmt::format(" - \"{}\"", S3_get_status_name((S3Status)status));
         }
 
-        result = ASSERT_ERROR(status == S3StatusOK, status, "[resource_name=%s] Error initializing the S3 library. Status = %d.", resource_name.c_str(),
-                              status, msg.str().c_str());
+        result = ASSERT_ERROR(status == S3StatusOK, status, msg);
```
Comment (korydraughn) [inline s3/libirods_s3.cpp:602]: We should avoid using these macros in new code / code changes.

File: s3/libirods_s3.cpp [hunk 4]
```diff
@@ -673,15 +675,15 @@ S3UriStyle s3_get_uri_request_style( irods::plugin_property_map& _prop_map)
...
+int64_t s3GetMaxUploadSizeMB (irods::plugin_property_map& _prop_map)
 {
     irods::error ret;
     std::string max_size_str;   // max size from context string, in MB
 
     ret = _prop_map.get<std::string>(s3_max_upload_size, max_size_str);
     if (ret.ok()) {
         // should be between 5MB and 5TB
-        long max_megs = atol(max_size_str.c_str());
+        int64_t max_megs = atol(max_size_str.c_str());
```
Comment (korydraughn) [inline s3/libirods_s3.cpp:686]: `atol` -> `std::atol` 

Include the `std` namespace for C functions as well.

File: s3/libirods_s3.cpp [hunk 5]
```diff
@@ -848,20 +849,22 @@ static void mrdWorkerThread (
...
-            msg.str( std::string() ); // Clear
-            msg << "Multirange:  Start range " << (int)seq << ", key \"" << g_mrdKey << "\", offset "
-                << (long)rangeData.get_object_data.offset << ", len " << (int)rangeData.get_object_data.contentLength;
-            rodsLog( LOG_DEBUG, msg.str().c_str() );
+            irods::log(LOG_DEBUG,
+                    fmt::format("Multirange:  Start range {}  \"{}\", offset {}, len {}",
+                    (int)seq,
+                    g_mrdKey,
+                    (int64_t)rangeData.get_object_data.offset,
+                    (int)rangeData.get_object_data.contentLength));
```
Comment (korydraughn) [inline s3/libirods_s3.cpp:857]: Are the C-style casts necessary here?

fmtlib deduces the correct types for you.

File: s3/libirods_s3.cpp [hunk 6]
```diff
@@ -937,7 +944,21 @@ size_t get_retry_wait_time_sec(irods::plugin_property_map& _prop_map) {
...
                 "[resource_name=%s] failed to cast %s [%s] to a size_t", resource_name.c_str(),
-                s3_wait_time_sec.c_str(), wait_time_str.c_str() );
+                s3_wait_time_seconds.c_str(), wait_time_str.c_str() );
+        }
+    } else {
+        // for backward compatibility, look for S3_WAIT_TIME_SEC
+        irods::error ret = _prop_map.get< std::string >( s3_wait_time_sec, wait_time_str );
+        if( ret.ok() ) {
+            try {
+                retry_wait = boost::lexical_cast<size_t>( wait_time_str );
```
Comment (korydraughn) [inline s3/libirods_s3.cpp:954]: `size_t` -> `std::size_t`

File: s3/libirods_s3.cpp [hunk 7]
```diff
@@ -1491,13 +1530,19 @@ static void mpuWorkerThread (
...
+                    seq);
+
             if(partData.status >= 0) {
-                msg << " - \"" << S3_get_status_name(partData.status) << "\"";
+                msg += fmt::format(" - \"{}\"", S3_get_status_name((S3Status)partData.status));
             }
-            result = ERROR( S3_PUT_ERROR, msg.str() );
-            rodsLog( LOG_ERROR, msg.str().c_str() );
+
+            result = ERROR( S3_PUT_ERROR, msg );
```
Comment (korydraughn) [inline s3/libirods_s3.cpp:1544]: The value of `result` can be assigned to `g_mpuResult` directly.

Not required by any means.

File: s3/libirods_s3.hpp [hunk 8]
```diff
@@ -47,8 +47,8 @@ extern const std::string  s3_circular_buffer_size;
 extern const std::string  s3_circular_buffer_timeout_seconds; // timeout for read or write to circular buffer
 extern const std::string  s3_uri_request_style;        //  either "path" or "virtual_hosted" - default "path"
 extern const std::string  s3_number_of_threads;        //  to save number of threads
-extern const size_t       S3_DEFAULT_RETRY_WAIT_SEC;
-extern const size_t       S3_DEFAULT_MAX_RETRY_WAIT_SEC;
+extern const size_t       S3_DEFAULT_RETRY_WAIT_SECONDS;
```
Comment (korydraughn) [inline s3/libirods_s3.hpp:50]: `size_t` -> `std::size_t`

File: s3/s3_transport/unit_tests/src/test_s3_transport.cpp [hunk 9]
```diff
@@ -59,138 +61,112 @@ void read_keys(const std::string& keyfile, std::string& access_key, std::string&
...
 
     // create the bucket
-    std::stringstream ss;
-    ss << "aws --endpoint-url http://" << hostname << " s3 mb s3://" << bucket_name_ss.str();
+    const auto aws_mb_command = fmt::format("aws --endpoint-url http://{} s3 mb s3://{}", hostname, bucket_name);
 
-    rodsLog(LOG_NOTICE, "%s\n", ss.str().c_str());
-    system(ss.str().c_str());
+    irods::log(LOG_NOTICE, aws_mb_command);
+    system(aws_mb_command.c_str());
```
Comment (korydraughn) [inline s3/s3_transport/unit_tests/src/test_s3_transport.cpp:72]: `system` -> `std::system`

**Review by korydraughn — COMMENTED** [review]
Looking good.

File: s3/libirods_s3.cpp [hunk 10]
```diff
@@ -122,8 +122,10 @@ const std::string  s3_auth_file{"S3_AUTH_FILE"};
 const std::string  s3_key_id{"S3_ACCESS_KEY_ID"};
 const std::string  s3_access_key{"S3_SECRET_ACCESS_KEY"};
 const std::string  s3_retry_count{"S3_RETRY_COUNT"};
-const std::string  s3_wait_time_sec{"S3_WAIT_TIME_SEC"};
-const std::string  s3_max_wait_time_sec{"S3_MAX_WAIT_TIME_SEC"};
+const std::string  s3_wait_time_seconds{"S3_WAIT_TIME_SECONDS"};
+const std::string  s3_wait_time_sec{"S3_WAIT_TIME_SEC"};                 // being deprecated
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:127]: Determined this is not possible as it add a deprecation failure on compile and don't want to disable deprecation failures.

File: s3/libirods_s3.cpp [hunk 11]
```diff
@@ -937,7 +939,21 @@ size_t get_retry_wait_time_sec(irods::plugin_property_map& _prop_map) {
             rodsLog(
                 LOG_ERROR,
                 "[resource_name=%s] failed to cast %s [%s] to a size_t", resource_name.c_str(),
-                s3_wait_time_sec.c_str(), wait_time_str.c_str() );
+                s3_wait_time_seconds.c_str(), wait_time_str.c_str() );
+        }
+    } else {
+        // for backward compatibility, look for S3_WAIT_TIME_SEC
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:950]: Added a warning message if the *SECONDS key is not found and the *SEC key is.

File: s3/libirods_s3.cpp [hunk 12]
```diff
@@ -957,10 +973,23 @@ size_t get_max_retry_wait_time_sec(irods::plugin_property_map& _prop_map) {
             rodsLog(
                 LOG_ERROR,
                 "[resource_name=%s] failed to cast %s [%s] to a size_t", resource_name.c_str(),
-                s3_max_wait_time_sec.c_str(), max_retry_wait_str.c_str() );
+                s3_max_wait_time_seconds.c_str(), max_retry_wait_str.c_str() );
+        }
+    } else {
+        // for backward compatibility, look for S3_MAX_WAIT_TIME_SEC
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:984]: Added a warning message if the *SECONDS key is not found and the *SEC key is.

File: s3/libirods_s3.cpp [hunk 13]
```diff
@@ -917,7 +917,7 @@ std::string get_cache_directory(irods::plugin_property_map& _prop_map) {
             s3_cache_dir_str = boost::filesystem::temp_directory_path().string();
         }
 
-        // No longer using shared_memory_name_salt because it uses irods main server process PID and a hash
+        // No int64_ter using shared_memory_name_salt because it uses irods main server process PID and a hash
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:925]: done

File: s3/libirods_s3.cpp [hunk 14]
```diff
@@ -2490,7 +2490,7 @@ class s3_resource : public irods::resource {
 
 
     // =-=-=-=-=-=-=-
-    // 3b. pass along a functor for maintenance work after
+    // 3b. pass aint64_t a functor for maintenance work after
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:2504]: Ugh.  done

File: s3/libirods_s3.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:110]: Going to change the commit message.  I will force put that change only.

File: s3/libirods_s3.cpp (hunk 3 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:602]: Understood but this will go in another pull request I believe.

File: s3/libirods_s3.cpp (hunk 4 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:686]: done for all atol

File: s3/libirods_s3.cpp (hunk 5 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:857]: I don't think so and I was sure you were going to ask.  ;-)

File: s3/libirods_s3.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply s3/libirods_s3.cpp:127]: We should log a message when the plugin sees this variable being used.

The message would be a warning that includes the deprecation notice.

File: s3/libirods_s3.cpp (hunk 5 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:857]: Done both here and in transport code.

File: s3/libirods_s3.cpp (hunk 7 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:1544]: done

File: s3/s3_transport/unit_tests/src/test_s3_transport.cpp [hunk 15]
```diff
@@ -16,8 +16,10 @@
 #include <chrono>
 #include <string>
 #include <sstream>
-
 #include <string_view>
+#include <fmt/format.h>
+#include "irods/miscServerFunct.hpp"
+#include "irods/data_object_finalize.h"
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/s3_transport/unit_tests/src/test_s3_transport.cpp:22]: Done.  The second one was not even required so it was removed.  I think it was an aggressive copy/paste.

File: s3/libirods_s3.cpp (hunk 1 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp]: dibe

File: s3/libirods_s3.cpp (hunk 6 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp]: Done

File: s3/libirods_s3.hpp (hunk 8 above)
Comment (JustinKyleJames) [reply s3/libirods_s3.hpp]: done

**Comment by JustinKyleJames** [comment]
I've made the suggested changes.  I did a force push only to have the one commit message updated.  The changes are in the last two commits.

File: s3/s3_transport/unit_tests/src/test_s3_transport.cpp [hunk 16]
```diff
@@ -18,8 +18,8 @@
 #include <sstream>
 #include <string_view>
 #include <fmt/format.h>
-#include "irods/miscServerFunct.hpp"
-#include "irods/data_object_finalize.h"
+#include <irods/miscServerFunct.hpp>
+//#include <irods/data_object_finalize.h>
```
Comment (alanking) [inline s3/s3_transport/unit_tests/src/test_s3_transport.cpp:22]: If it's not needed, we should just delete it.

**Review by alanking — APPROVED** [review]
Very nice. Awaiting approval from @korydraughn and then we can squash

File: s3/s3_transport/unit_tests/src/test_s3_transport.cpp (hunk 16 above)
Comment (korydraughn) [reply s3/s3_transport/unit_tests/src/test_s3_transport.cpp:22]: yes. please remove that line and squash it.

once that is done, i think we're good.

Comment (JustinKyleJames) [reply s3/s3_transport/unit_tests/src/test_s3_transport.cpp]: done

**Comment by alanking** [comment]
Sweet. Please squash commits, but leave off #'s for now so we can look at it first

**Review by korydraughn — COMMENTED** [review]
Everything looks good to me. I'd recommend rewording the deprecation warnings if those property names are indeed deprecated.

**Review by korydraughn — APPROVED** [review]

**Pull request merged**